### ISSUE TITLE
cmd/stdiscosrv: Not enough traffic breaks replication between discovery servers

### ISSUE THREAD

**Issue by vespian** [syncthing/syncthing#5117]
### Environment:

Two discovery servers, one running v0.14.46@armhf/Raspbian, the other running on v0.14.49@x86_64/Centos7. I occasionally see in the logs:
```
stdiscosrv[11974]: Replication read size: read tcp 172.31.3.1:19200->192.168.4.1:50316: i/o timeout
```
or
```
stdiscosrv[12505]: Replication write: write tcp 192.168.4.2:40726->192.168.8.1:19200: write:broken pipe
stdiscosrv[12505]: Supervisor main: main: Failed service 'replicationSender("192.168.8.1:19200")' (1.000000 failures of 5.000000), restarting: true, error: "{replicationSender(\"192.168.8.1:19200\") replicationSender(\"192.168.8.1:19200\")} returned unexpectedly", stacktrace: [unknown stack trace]
```
After closer investigation, it seem that the culprit is this line: https://github.com/syncthing/syncthing/blob/2c9d96375bb5e3a8dcf7840412462182ce0b9433/cmd/stdiscosrv/replication.go#L248-L252

Namely, unless there is enough traffic on the wire, the replication connection will be torn down. What is more, the is no reconnect on the other side: https://github.com/syncthing/syncthing/blob/2c9d96375bb5e3a8dcf7840412462182ce0b9433/cmd/stdiscosrv/replication.go#L105-L110

So basically, if there is not enough traffic in the replication servers, the replication will not work.

I think the solution would be to not to set a read timeout in https://github.com/syncthing/syncthing/blob/2c9d96375bb5e3a8dcf7840412462182ce0b9433/cmd/stdiscosrv/replication.go#L245 - i.e. wait indefinitely (preferable), or handle reconnects on the sending side (more work), or implement heartbeats in the replication protocol (quite a bit of work).

Let me know if this sounds reasonable. I can prepare a patch.

### Version Information
```
stdiscosrv v0.14.46-ds1 (go1.10rc2 linux-arm)
stdiscosrv v0.14.49 (go1.9.4 linux-amd64)
```

**Comment by AudriusButkevicius** [comment]
I think the right approach would be to send heartbeats, yet I'll leave for @calmh to confirm.

**Comment by calmh** [comment]
That said, it does reconnect, so this is a cosmetic problem for low traffic setups only.

**Comment by calmh** [comment]
You're right, one replication message is sacrificed on the sending side to discover the lost connection, then it reconnects.

**Comment by vespian** [comment]
@calmh  Attempt number 1, please have a look:

https://github.com/syncthing/syncthing/pull/5120/files

**Comment by vespian** [comment]
I was thining a bit on how we can implement re-sending the event in case when replication connection turns out to be broken, but this would take more work as the replication events should probably be ordered - we can't just re-add them onto the end of the queue/channel. Right?
